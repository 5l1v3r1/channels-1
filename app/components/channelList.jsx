// @flow
import React from 'react';
import Channel from './channel';
import Loader from './loader';
import BlankState from './blankState';
import ErrorState from './errorState';
import ChannelType from '../dataTypes/channelType';
import styles from './channelList.scss';

const ChannelList = ({
  channels,
  isFetching,
  error,
  isOpen,
  fetchChannels,
  selectChannel,
  toggleChannelList
}: {
  channels: ChannelType[],
  isFetching: boolean,
  error: boolean,
  isOpen: boolean,
  fetchChannels: () => void,
  selectChannel: (id: string) => void,
  toggleChannelList: () => void
}) => {
  if (isFetching || error) {
    return (
      <div className={isOpen ? `${styles.channelList} ${styles['channelList--isOpen']}` : styles.channelList}>
        <div className={styles.closeTarget} onClick={toggleChannelList} />
        <div className={styles.modal}>
          {isFetching ? (
            <div className={styles.loader}>
              <Loader />
            </div>
          ) : (
            <div className={styles.errorState}>
              <ErrorState
                message={'Error requesting channels.'}
                retry={fetchChannels}
              />
            </div>
          )}
        </div>
      </div>
    );
  }

  return (
    <div className={isOpen ? `${styles.channelList} ${styles['channelList--isOpen']}` : styles.channelList}>
      <div className={styles.closeTarget} onClick={toggleChannelList} />
      <div className={styles.modal}>
        {channels.length ? (
          <ul className={styles.channels}>
            {channels.map(channel =>
              <Channel
                key={channel.id}
                title={channel.title}
                thumbnail={channel.thumbnail}
                videoCount={channel.videoCount}
                subscriberCount={channel.subscriberCount}
                isActive={channel.isActive}
                selectChannel={() => selectChannel(channel.id)}
              />
            )}
          </ul>
        ) : (
          <div className={styles.blankState}>
            <BlankState
              message={'No channels found.'}
            />
          </div>
        )}
      </div>
    </div>
  );
}

export default ChannelList;
