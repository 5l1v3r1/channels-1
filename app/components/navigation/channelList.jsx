// @flow
import React from 'react';
import Channel from './channel';
import Loader from '../loader';
import BlankState from '../blankState';
import ErrorState from '../errorState';
import ChannelType from '../../dataModels/channel';
import styles from './channelList.scss';

const ChannelList = ({
  channels,
  isFetching,
  error,
  isOpen,
  fetchChannels,
  selectChannel,
  toggleVisibility
}: {
  channels: ChannelType[],
  isFetching: boolean,
  error: boolean,
  isOpen: boolean,
  fetchChannels: () => void,
  selectChannel: (id: string) => void,
  toggleVisibility: () => void
}) => {
  if (isFetching || error) {
    return (
      <div className={isOpen ? `${styles.channelList} ${styles['channelList--isOpen']}` : styles.channelList}>
        <div className={styles.closeTarget} onClick={toggleVisibility} />
        <div className={styles.modal}>
          {isFetching ? (
            <div className={styles.loader}>
              <Loader />
            </div>
          ) : (
            <div className={styles.errorState}>
              <ErrorState
                message={'Error requesting channels.'}
                retry={fetchChannels}
              />
            </div>
          )}
        </div>
      </div>
    );
  }

  return (
    <div className={isOpen ? `${styles.channelList} ${styles['channelList--isOpen']}` : styles.channelList}>
      <div className={styles.closeTarget} onClick={toggleVisibility} />
      <div className={styles.modal}>
        {channels.length ? (
          <ul className={styles.channels}>
            {channels.map(channel =>
              <Channel
                key={channel.id}
                title={channel.title}
                thumbnail={channel.thumbnail}
                videoCount={channel.videoCount}
                subscriberCount={channel.subscriberCount}
                isActive={channel.isActive}
                selectChannel={() => selectChannel(channel.id)}
              />
            )}
          </ul>
        ) : (
          <div className={styles.blankState}>
            <BlankState
              message={'No channels found.'}
            />
          </div>
        )}
      </div>
    </div>
  );
}

export default ChannelList;
