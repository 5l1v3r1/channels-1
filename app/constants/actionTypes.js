// @flow
const actionTypes = {
  REQUEST_ACCESS_TOKEN: 'REQUEST_ACCESS_TOKEN',
  RECEIVE_ACCESS_TOKEN: 'RECEIVE_ACCESS_TOKEN',
  ACCESS_TOKEN_ERROR: 'ACCESS_TOKEN_ERROR',
  REQUEST_CHANNELS: 'REQUEST_CHANNELS',
  RECEIVE_CHANNELS: 'RECEIVE_CHANNELS',
  CHANNELS_ERROR: 'CHANNELS_ERROR',
  TOGGLE_CHANNEL_LIST: 'TOGGLE_CHANNEL_LIST',
  UPDATE_ACTIVE_CHANNEL: 'UPDATE_ACTIVE_CHANNEL',
  UPDATE_STATUS: 'UPDATE_STATUS',
  ENABLE_SEARCH_BAR: 'ENABLE_SEARCH_BAR'
};

export default actionTypes;
