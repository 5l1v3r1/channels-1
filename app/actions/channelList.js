// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch, AccessTokenState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

type GetState = () => { accessToken: AccessTokenState };

type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { accessToken } = getState().accessToken;
    const subscriptions = fetchSubscriptions(accessToken);
    subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(accessToken, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          dispatch(receiveChannels(json));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

function fetchSubscriptions(accessToken: string): Promise<any> {
  const uri = getSubscriptionsUri(accessToken);
  return apiRequest(uri, Methods.GET);
}

function requestChannels(): Action {
  return {
    type: actionTypes.REQUEST_CHANNELS
  };
}

function receiveChannels(json: any): Action {
  return {
    type: actionTypes.RECEIVE_CHANNELS,
    payload: json.items
  };
}

function channelsError(): Action {
  return {
    type: actionTypes.CHANNELS_ERROR
  };
}
