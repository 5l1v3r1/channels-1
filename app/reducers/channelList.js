// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ChannelListState as State } from '../constants/typeAliases';
import Channel from '../dataModels/channel';
import channelReducer from './channel';

const initialState = {
  channels: [],
  isFetching: false,
  error: false,
  isOpen: false
};

export default function channelList(state: State = initialState, { type, payload }: Action): State {
  switch (type) {
    case actionTypes.REQUEST_CHANNELS:
      return {
        ...state,
        isFetching: true,
        error: false
      };
    case actionTypes.RECEIVE_CHANNELS:
      return {
        ...state,
        channels: !payload ? [] : payload.map(json => Channel.from(json)),
        isFetching: false,
        error: false
      };
    case actionTypes.CHANNELS_ERROR:
      return {
        ...state,
        isFetching: false,
        error: true
      };
    case actionTypes.TOGGLE_VISIBILITY:
      return {
        ...state,
        isOpen: !state.isOpen
      };
    case actionTypes.SELECT_CHANNEL:
      return {
        ...state,
        channels: state.channels.map(channel => channelReducer(channel, { type, payload }))
      };
    default:
      return state;
  }
}
