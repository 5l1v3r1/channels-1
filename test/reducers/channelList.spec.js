import channelList from '../../app/reducers/channelList';
import actionTypes from '../../app/constants/actionTypes';
import ChannelType from '../../app/dataTypes/channelType';

describe('channelList reducer', () => {
  it('should return the initial state', () => {
    expect(channelList(undefined, {})).toMatchSnapshot();
  });

  it('should handle REQUEST_CHANNELS', () => {
    const action = { type: actionTypes.REQUEST_CHANNELS };
    expect(channelList(undefined, action)).toMatchSnapshot();
  });

  it('should handle RECEIVE_CHANNELS', () => {
    ChannelType.from = jest.fn().mockImplementation(json => new ChannelType(json.title));

    const stateBefore = channelList(undefined, { type: actionTypes.REQUEST_CHANNELS });
    const action = {
      type: actionTypes.RECEIVE_CHANNELS,
      payload: [
        { title: 'DevTips' },
        { title: 'Fun Fun Function' }
      ]
    };
    const stateAfter = channelList(stateBefore, action);

    expect(stateAfter.channels.length).toBe(action.payload.length);
  });

  it('should handle CHANNELS_ERROR', () => {
    const state = channelList(undefined, { type: actionTypes.REQUEST_CHANNELS });
    const action = { type: actionTypes.CHANNELS_ERROR };
    expect(channelList(state, action)).toMatchSnapshot();
  });

  it('should handle TOGGLE_CHANNEL_LIST', () => {
    const action = { type: actionTypes.TOGGLE_CHANNEL_LIST };
    expect(channelList(undefined, action)).toMatchSnapshot();
  });

  it('should handle UPDATE_ACTIVE_CHANNEL', () => {
    const state = {
      channels: [
        new ChannelType(
          'DevTips',
          'https://yt3.ggpht.com/photo.jpg',
          '292',
          '232937',
          false,
          '8ef09d6c-1eda-4e1e-aeec-e3ddff47dd45'
        ),
        new ChannelType(
          'Fun Fun Function',
          'https://yt3.ggpht.com/photo.jpg',
          '114',
          '98801',
          false,
          'bc91ca60-42dd-4694-bfbf-88edf8ca2805'
        )
      ],
      isFetching: false,
      error: false,
      isOpen: false
    };
    const action = {
      type: actionTypes.UPDATE_ACTIVE_CHANNEL,
      payload: 'bc91ca60-42dd-4694-bfbf-88edf8ca2805'
    };
    expect(channelList(state, action)).toMatchSnapshot();
  });

  it('should handle unknown action type', () => {
    const action = { type: 'unknown' };
    expect(channelList(undefined, action)).toMatchSnapshot();
  });
});
