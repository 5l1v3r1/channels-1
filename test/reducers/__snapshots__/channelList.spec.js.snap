// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`channelList reducer should handle CHANNELS_ERROR 1`] = `
Object {
  "channels": Array [],
  "error": true,
  "isFetching": false,
  "isOpen": false,
}
`;

exports[`channelList reducer should handle REQUEST_CHANNELS 1`] = `
Object {
  "channels": Array [],
  "error": false,
  "isFetching": true,
  "isOpen": false,
}
`;

exports[`channelList reducer should handle TOGGLE_CHANNEL_LIST 1`] = `
Object {
  "channels": Array [],
  "error": false,
  "isFetching": false,
  "isOpen": true,
}
`;

exports[`channelList reducer should handle UPDATE_ACTIVE_CHANNEL 1`] = `
Object {
  "channels": Array [
    ChannelType {
      "id": "8ef09d6c-1eda-4e1e-aeec-e3ddff47dd45",
      "isActive": false,
      "subscriberCount": "232937",
      "thumbnail": "https://yt3.ggpht.com/photo.jpg",
      "title": "DevTips",
      "videoCount": "292",
    },
    ChannelType {
      "id": "bc91ca60-42dd-4694-bfbf-88edf8ca2805",
      "isActive": true,
      "subscriberCount": "98801",
      "thumbnail": "https://yt3.ggpht.com/photo.jpg",
      "title": "Fun Fun Function",
      "videoCount": "114",
    },
  ],
  "error": false,
  "isFetching": false,
  "isOpen": false,
}
`;

exports[`channelList reducer should handle unknown action type 1`] = `
Object {
  "channels": Array [],
  "error": false,
  "isFetching": false,
  "isOpen": false,
}
`;

exports[`channelList reducer should return the initial state 1`] = `
Object {
  "channels": Array [],
  "error": false,
  "isFetching": false,
  "isOpen": false,
}
`;
